### ISSUE TITLE
 xcode 9.3 archive crash in FMDatabaseQueue

### ISSUE THREAD

**Issue by echooooooo** [ccgus/fmdb#659]
I found a problem with the xcode 9.3 system. I use xcode9.3 to run my app with no problem. However, after using xcode to generate the ad hoc installation package, the app will crash. The location of the problem is a tripartite library fmdb. Then the same code, I use xcode9.2 to generate ad hoc installation package, everything is normal after installation, without any crash. I do not know where xcode 9.3 has been modified, resulting in the use of xcode 9.3 unison to run normally does not crash, but the installation package installation will crash. Is the use of fmdb wrong or that this is a problem with xcode9.3. Trouble give me an answer.

**Comment by mheist09** [comment]
We were getting same error, worked fine in dev and release builds until xcode 9.3.

we were recreating a database queue that was already open:

database_queue = [FMDatabaseQueue databaseQueueWithPath:path];

so dealloc was getting called on existing instance, that instance maybe was processing transactions but think it would crash even if it was idle.

I got it to stop crashing by not recreating the queue if it already existed, didn't try closing the queue before recreating but that might work too.

**Comment by johnryan** [comment]
@mheist09 interesting...I *think* my code should be immune to that already...i have a singleton DatabaseManager with a `queue` method:

```
- (FMDatabaseQueue*)queue{
    @synchronized(self) {
        if (!_queue) {
            self.queue = [FMDatabaseQueue databaseQueueWithPath:path];
            [_queue inDatabase:^(FMDatabase *db) {
                //some extra db setup
            }]
        }
        return _queue;
    }
}
```

**Comment by mheist09** [comment]
I had to put a breakpoint on FMDatabaseQueue dealloc to figure out when it was getting called, the code will run the same in debug but only crash in production. Maybe yours is only happening when the app is shutting down?

**Comment by johnryan** [comment]
I don't think so, based on the logs I have in my crash reports it appears to happen on startup. I'll have to stare at this a bit more to think if there are any race conditions possible
